MITgcm/MITgcm
# OpenAD DIVA

## What changes does this PR introduce?
(Bug fix, feature, docs update, ...)
It implements divided adjoints for packages differentiated by OpenAD. Both serial and parallel runs are supported. 


## What is the current behaviour? 
(You can also link to an open issue here)
Divided adjoints using OpenAD is currently broken. 

## What is the new behaviour 
(if this is a feature change)?
When guided by a script, the adjoint computation can suspend itself after a specified number of iterations. The computation can then be restarted using previously written checkpoints.


## Does this PR introduce a breaking change? 
(What changes might users need to make in their application due to this PR?)
No. 

## Other information:

How to run a serial verification test using the diva capability (assuming that one can already run OpenAD verification tests):
1. ` mkdir MITgcm/verification/TESTNAME/build/`
2.  `cd MITgcm/verification/TESTNAME/build/`
3.  Add the following line to OPENAD_OPTIONS.h in the code mods folder.
`#define ALLOW_OPENAD_DIVA`
4. Invoke genmake2 using diva option
`../../../tools/genmake2 -oad -mods [FOLDER_WHERE_CODE_MODS_RESIDE] -ieee -diva --rootdir=[FOLDER_OF_MITGCM_CODE]`
5. `make adAll`

Once compile is successful, one can run the code:
1.  `cd ../run`
2. `ln -s ../build/mitgcmuv_ad .`
3. ` ln -s ../input_oad/* .`
4. Change the values of nTimeSteps and nTimeSteps_l2 in the 'data' file. If the value of nTimesSteps without DIVA is x, then for the DIVA case, the product nTimeSteps*nTimeSteps_l2 must equal x
5. Edit the `prepare diva` script to change how often adjoint checkpoints should take place and what the starting iteration is. 
6. Invoke the `prepare diva` script which creates a file named `oad_reg_state.000`. 
7.  Place `oad_reg_state.000` in `run` folder
8.  Run model multiple times as needed. For example:
`./mitgcmuv_ad >out1.txt`
`STOP DIVA FINISHED FORWARD`
`./mitgcmuv_ad >out2.txt`
`STOP DIVA SUSPEND ADJOINT`
`./mitgcmuv_ad >out3.txt`
STOP NORMAL END


## Suggested addition to `tag-index`
(To avoid unnecessary merge conflicts, please don't update `tag-index`. One of the admins will do that when merging your pull request.)
